File: src/haruhiloop_cli/mutator.py
from __future__ import annotations

import random
from dataclasses import dataclass
from typing import Protocol

from haruhiloop_cli.models import GameState


PROFILE_KEYS = ("satisfaction_factor", "stability_factor", "clue_factor")


@dataclass(slots=True)
class MutationProfile:
    satisfaction_factor: float = 1.0
    stability_factor: float = 1.0
    clue_factor: float = 1.0

    def to_dict(self) -> dict[str, float]:
        return {
            "satisfaction_factor": self.satisfaction_factor,
            "stability_factor": self.stability_factor,
            "clue_factor": self.clue_factor,
        }


class WorldlineMutator(Protocol):
    def mutate(self, state: GameState) -> MutationProfile: ...


class DeterministicMutator:
    def mutate(self, state: GameState) -> MutationProfile:
        # Deterministic profile based only on persisted state.
        loop_bias = min(0.08, state.loop_count * 0.005)
        stability_bias = 0.05 if state.stability <= 40 else 0.0
        clue_bias = min(0.1, state.memory_residue.get("clue_efficiency", 0) * 0.02)
        return MutationProfile(
            satisfaction_factor=1.0 - loop_bias,
            stability_factor=1.0 - stability_bias,
            clue_factor=1.0 + clue_bias,
        )


class AIMutator:
    def __init__(self, seed: int | None = None, temperature: float = 0.7) -> None:
        self._rng = random.Random(seed)
        self._temperature = max(0.0, temperature)

    def mutate(self, state: GameState) -> MutationProfile:
        spread = max(0.02, min(0.25, self._temperature * 0.2))
        raw = MutationProfile(
            satisfaction_factor=1.0 + self._rng.uniform(-spread, spread),
            stability_factor=1.0 + self._rng.uniform(-spread, spread),
            clue_factor=1.0 + self._rng.uniform(-spread, spread),
        )
        if state.closed_space_stage > 0:
            raw.stability_factor -= 0.03
        if state.crew_sync >= 65:
            raw.satisfaction_factor += 0.02
        return raw


def validate_profile(profile: MutationProfile | dict[str, float]) -> dict[str, float]:
    if isinstance(profile, MutationProfile):
        payload = profile.to_dict()
    else:
        payload = dict(profile)
    validated: dict[str, float] = {}
    for key in PROFILE_KEYS:
        value = float(payload.get(key, 1.0))
        validated[key] = max(0.8, min(1.2, value))
    return validated


def build_mutator(mode: str, seed: int | None = None, temperature: float = 0.7) -> WorldlineMutator:
    if mode == "deterministic":
        return DeterministicMutator()
    if mode == "ai":
        return AIMutator(seed=seed, temperature=temperature)
    raise ValueError(f"未知扰动模式：{mode}")

File: web/src/domain/mutator.ts
// 与 src/haruhiloop_cli/mutator.py 1:1 对齐（Deterministic 公式相同；
// AI 模式用 mulberry32 替代 Python random.Random，逻辑等价但二进制不一致）。

import { GameState, MutationProfile } from "./models";
import { mulberry32, randUniform } from "../infrastructure/prng";

export const PROFILE_KEYS = [
  "satisfaction_factor",
  "stability_factor",
  "clue_factor",
] as const;

export interface WorldlineMutator {
  mutate(state: GameState): MutationProfile;
}

export class DeterministicMutator implements WorldlineMutator {
  mutate(state: GameState): MutationProfile {
    const loopBias = Math.min(0.08, state.loop_count * 0.005);
    const stabilityBias = state.stability <= 40 ? 0.05 : 0.0;
    const clueBias = Math.min(0.1, (state.memory_residue["clue_efficiency"] ?? 0) * 0.02);
    return {
      satisfaction_factor: 1.0 - loopBias,
      stability_factor: 1.0 - stabilityBias,
      clue_factor: 1.0 + clueBias,
    };
  }
}

export class AIMutator implements WorldlineMutator {
  private readonly rng: () => number;
  private readonly temperature: number;

  constructor(seed: number | null = null, temperature = 0.7) {
    // 在没有 seed 时也保留可重现性：以当前时间戳作为 entropy。
    const seedNum = seed ?? ((Date.now() & 0xffffffff) >>> 0);
    this.rng = mulberry32(seedNum);
    this.temperature = Math.max(0.0, temperature);
  }

  mutate(state: GameState): MutationProfile {
    const spread = Math.max(0.02, Math.min(0.25, this.temperature * 0.2));
    const raw: MutationProfile = {
      satisfaction_factor: 1.0 + randUniform(this.rng, -spread, spread),
      stability_factor: 1.0 + randUniform(this.rng, -spread, spread),
      clue_factor: 1.0 + randUniform(this.rng, -spread, spread),
    };
    if (state.closed_space_stage > 0) raw.stability_factor -= 0.03;
    if (state.crew_sync >= 65) raw.satisfaction_factor += 0.02;
    return raw;
  }
}

export function validateProfile(profile: MutationProfile): MutationProfile {
  return {
    satisfaction_factor: Math.max(0.8, Math.min(1.2, profile.satisfaction_factor)),
    stability_factor: Math.max(0.8, Math.min(1.2, profile.stability_factor)),
    clue_factor: Math.max(0.8, Math.min(1.2, profile.clue_factor)),
  };
}

export function buildMutator(
  mode: string,
  seed: number | null = null,
  temperature = 0.7,
): WorldlineMutator {
  if (mode === "deterministic") return new DeterministicMutator();
  if (mode === "ai") return new AIMutator(seed, temperature);
  throw new Error(`未知扰动模式：${mode}`);
}
